### ISSUE TITLE
Multiple applications, duplicate events.

### ISSUE THREAD

**Issue by Singletoned** [johnbywater/eventsourcing#135]
I've been reading through the documentation and I'm quite keen to try using the library but one thing caught my eye, which was the references to having to be careful only to instantiate one copy of the application.

> "If your eventsourcing application object has any policies, for example if is has a persistence policy that will persist events whenever they are published, then constructing more than one instance of the application causes the policy event handlers to be subscribed more than once, so for example more than one attempt will be made to save each event, which won’t work."

> "When deploying an event sourcing application with Django, just remember that there must only be one instance of the application in any given process, otherwise its subscribers will be registered too many times."

Is this duplication in publish and subscribe the only problem that you are aware of?

From a superficial reading of the code it looks as it would be possible to fix that, as you are using a global `subscribe` method and a global `publish` method, which are both hooked to a global store of event_handlers.  Could these be moved to the Application object?  Or is there an architectural reason for having it this way?

It certainly feels like it would make things cleaner and easier by not having these globals, but I realise I may be missing something from the bigger picture (I'm not deeply experienced with Event Sourcing in general).

I'd be happy to make the changes and do a PR if you think its a good idea.

Thanks

Ed

**Comment by johnbywater** [comment]
Thanks Ed. It would be good to make this aspect clearer. Although I also feel there is something that remains unresolved, I'm not exactly sure what it is.

Regarding your suggestion to move the global store of event_handlers from a module-level variable to an attribute of the application object: the domain entity objects need to publish events somehow, and by design the domain layer doesn't have a dependency on the application layer or the infrastructure layer. By design, you can develop and test the domain entity object behaviour as a normal Python object, without any dependencies, without needing a surrounding framework to make anything work. But it remains unavoidable that the persistence of entity state depends on persistence infrastructure. The pub-sub pattern is used to decouple this dependency, to invert control so that the model doesn't depend on a framework. The residual from this inversion of control (the dependency that remains after that dependency is removed) is the dependency of the domain model on the "pub" side of a pub-sub mechanism.

The answer to your first question is Yes. The documentation suggests that only one instance of "the application" class should exist in any given process. The reason given by the documentation is that otherwise the subscriptions of an application could be duplicated, and if e.g. database records that can only be inserted once are inserted twice: the second attempt will fail and your Python program will malfunction. However, it wouldn't be a problem for two applications with different subscriptions to exist in the same process: either if "the  application" class were replaced with two classes, each having half of the original responsibilities; or if an application class is parameterized, so that different instances effectively have different subscriptions. As you correctly identified, the important thing is to avoid duplicating subscriptions mistakenly.

I tried to make it explicit in the documentation, that duplicating subscriptions can be avoided by making sure you only have one instance of the application, which is anyway all you need to support the domain layer with infrastructure and provide something that interfaces can use. The library application classes are effectively configuration of the infrastructure for the model, and in general one should only need to configure things once. Hopefully the documentation describes the current situation, but it's awkward to present these application object classes, and then disclose that everything is pegged to a global variable so you can only have one instance.

The existing library application classes are effectively singletons, so perhaps they could be coded as such, in a way that supports subclassing. The complications discussed in the documentation (amongst the ``init_application`` vs ``get__application`` vs ``close__application`` functions) could probably all be moved behind a suitable singleton interface, a lock could be used to avoid the race condition around the construction of the singleton in requests, and a close() method could be implemented that ends by dropping a reference to the single instance, so that the lifecycle is closed and we can go around that loop. Some of the documentation could then perhaps be deleted, and things would appear cleaner?

Looking around a bit, despite the disdain for the singleton pattern, it is quite easy to find singletons in popular Python libraries. I think perhaps part of me wanted to avoid saying "singleton" to avoid having to listen to people complaining that they don't like singletons. There are lots of singletons in Python libraries. The actor frameworks I've been looking at all have a singleton "actor system" object, often coded as a singleton which probably helps to make things clear. The actor system object of the Thespian library needs to be "shutdown", which is equivalent to the ``close()`` method of the application classes in the library. So there's no reason why the library's application classes couldn't be recoded as singletons that can be similarly shutdown and afterwards recreated.

Another thing that feels like a singleton is the scoped session facade in SQLAlchemy. It is able to access a concrete session object in different ways: thread-scoping uses the thread identity to select a concrete session object; request-scoping examines the call stack to find a request. So perhaps we could change the pub-sub mechanism to work like that. If an application subscribes handlers, then the subscriptions could be scoped to the application, and when calls to the application services invoke commands on entities that publish events, the pub-sub mechanism could somehow detect the application from the call stack, and use only the subscriptions that are scoped to that application. An alternative could be to scope the subscriptions to a request, and construct a new application object for each request, so that handlers are subscribed for a request scoped to that request, and used if events are published during that request. Then different applications could be used in the same application server without interference.

Searching the call stack to detect the scope of event handlers seems relatively complicated and possibly time consuming. When would somebody want to do that? I guess there could be an operational efficiency in having many different eventsourcing applications being served by the same application server layer. Constructing the application object for each request would be relatively inefficient compared with doing it once per process. But such inefficiencies might be marginal, I haven't timed how long it takes to construct an application object, and I haven't coded anything that scopes handler subscriptions.

So in summary, if you're asking me what I think about this, I suspect that without more justification, the complexity of implementing scoped event handling suggests that reworking the library application classes as singletons would most easily allow the library documentation and users' code to be simplified?

What do you think? :-)
